src/parsec.rs

```diff
@@ -15,41 +15,118 @@ use crate::{
 use rand::Rng;
-use std::collections::{btree_map::Entry, BTreeMap};
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    fmt,
+};
 
 #[cfg(feature = "mock_parsec")]
 pub use crate::mock::parsec::{
-    init_mock, ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId,
+    init_mock, ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId,
 };
 #[cfg(not(feature = "mock_parsec"))]
-pub use parsec::{ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId};
+pub use parsec::{ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId};
 
 pub type Block = inner::Block<chain::NetworkEvent, id::PublicId>;
 pub type Parsec = inner::Parsec<chain::NetworkEvent, FullId>;
 pub type Request = inner::Request<chain::NetworkEvent, id::PublicId>;
 pub type Response = inner::Response<chain::NetworkEvent, id::PublicId>;
 
+// TODO: Needs to be determined from better data.
+#[cfg(not(feature = "mock_parsec"))]
+const PARSEC_SIZE_LIMIT: u64 = 1_000_000_000;
+// Mock parsec request/responses are much smaller, so we need a lower limit.
+#[cfg(feature = "mock_parsec")]
+const PARSEC_SIZE_LIMIT: u64 = 100;
+
+// Keep track of size in case we need to prune.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct ParsecSizeCounter(u64);
+
+impl ParsecSizeCounter {
+    fn increase_size(&mut self, size: u64) {
+        self.0 += size;
+    }
+
+    fn needs_pruning(&self) -> bool {
+        self.0 > PARSEC_SIZE_LIMIT
+    }
+}
+
+impl fmt::Display for ParsecSizeCounter {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "size: {}", self.0)
+    }
+}
+
+struct ParsecAndCounter {
+    pub parsec: Parsec,
+    pub size_counter: ParsecSizeCounter,
+}
+
+impl ParsecAndCounter {
+    fn new(full_id: FullId, gen_pfx_info: &GenesisPfxInfo) -> Self {
+        let parsec = create(full_id, gen_pfx_info);
+        let size_counter = Default::default();
+
+        Self {
+            parsec,
+            size_counter,
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        src: &id::PublicId,
+        req: Request,
+        log_ident: &LogIdent,
+    ) -> Result<Response, Error> {
+        self.count_size(serialisation::serialised_size(&req), log_ident);
+        self.parsec.handle_request(src, req)
+    }
+
+    fn handle_response(
+        &mut self,
+        src: &id::PublicId,
+        resp: Response,
+        log_ident: &LogIdent,
+    ) -> Result<(), Error> {
+        self.count_size(serialisation::serialised_size(&resp), log_ident);
+        self.parsec.handle_response(src, resp)
+    }
+
+    fn count_size(&mut self, size: u64, log_ident: &LogIdent) {
+        self.size_counter.increase_size(size);
+        trace!(
+            "{} - Parsec size is now estimated to: {}.",
+            log_ident,
+            self.size_counter,
+        );
+    }
+}
+
 pub struct ParsecMap {
-    map: BTreeMap<u64, Parsec>,
+    map: BTreeMap<u64, ParsecAndCounter>,
```

Comment:
Sorry for that coming so late in the design, do we need `ParsecAndCounter`?
why is it not 
pub struct ParsecMap {
   map: BTreeMap<u64, Parsec>,
   size_counter: ParsecSizeCounter,
}

Essentially we reset it to 0 when a new parsec is added, and increment it only if it is the last parsec instance.
We are already doing special thing in this case `let poll = self.last_version() == msg_version;` It seem it would be reasonable to add an if condition, or something similar.

Reply:
Well it was added mostly as a response to previous review comments: "A struct for (parsec, size_counter) would probably allow you to avoid duplication here."

I'm still of the opinion that keeping a counter per parsec seems the cleanest option. Especially since it sounded like we might end up pruning frequently. Some of the code that uses this though was removed in an earlier revision though due to review comments

Reply:
> Well it was added mostly as a response to previous review comments: "A struct for (parsec, size_counter) would probably allow you to avoid duplication here."

True, that comment made sense in the context of we want counter per parsec, that new comment is taking a step back, and so make that less relevant.

> I'm still of the opinion that keeping a counter per parsec seems the cleanest option. 

Why, what is the purpose of the old counters?

Reply:
> Especially since it sounded like we might end up pruning frequently.
Why would that matter?

> Some of the code that uses this though was removed in an earlier revision though due to review comments
Yes, a contributing factor in me not questioning that design initially.

Reply:
> Why, what is the purpose of the old counters?

I mean we add to all parsecs, so doesn't it make sense to keep track how much we add to all of them?

It seems to me that for example we could use that to keep adding to the older parsecs up until they are full at which we use the more strict response suggested earlier (e.g. not replying at all)

Reply:
> It seems to me that for example we could use that to keep adding to the older parsecs up until they are full at which we use the more strict response suggested earlier (e.g. not replying at all)

But why not just do that immediately when it become not the last, as oppose to have that intermediate state where for some node that have reach size and do something different than the other nodes? (solution in search of a problem?)